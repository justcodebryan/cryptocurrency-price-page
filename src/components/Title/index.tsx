import classNames from 'classnames'
import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren, cloneElement } from 'react'

import styles from './styles.module.scss'

type TitleProps = PropsWithChildren<
  DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement> & {
    level?: 1 | 2 | 3 | 4
  }
>

const Title = ({ level, style, className, children, ...rest }: TitleProps) => {
  let renderItem

  const internalClass = classNames([styles['title'], level ? styles[`title-${level}`] : '', className])

  switch (level) {
    case 1:
      renderItem = <h1>{children}</h1>
      break
    case 2:
      renderItem = <h2>{children}</h2>
      break
    case 3:
      renderItem = <h3>{children}</h3>
      break
    case 4:
      renderItem = <h4>{children}</h4>
      break
    default:
      renderItem = <div>{children}</div>
  }

  return cloneElement(renderItem, {
    ...rest,
    style,
    className: internalClass,
  })
}

export default Title
