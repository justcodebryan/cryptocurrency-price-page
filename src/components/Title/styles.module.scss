.title {
  width: 100%;
  line-height: 1.5;

  &-1 {
    font-size: 36px;
    font-weight: 700;
    line-height: 1.2;
  }

  &-2 {
    font-size: 24px;
    font-weight: 600;
  }

  &-3 {
    font-size: 20px;
    font-weight: 600;
  }

  &-4 {
    font-size: 16px;
  }
}