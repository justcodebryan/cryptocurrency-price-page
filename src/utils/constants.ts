import { NoopFn } from '@/types'

export const NOOP: NoopFn = () => void 0

export const DEFAULT_MODAL_CONTAINER_ID = 'modal-container'

export const DEFAULT_POLLING_INTERVAL = 10 * 1000
