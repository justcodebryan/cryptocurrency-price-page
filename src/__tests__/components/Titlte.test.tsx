import Title from '@/components/Title'
import { toJSON } from '@/utils/test-utils'
import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'

describe('Title Test Suit', () => {
  test('Title Snapshot', () => {
    const component = renderer.create(<Title />)
    const tree = toJSON(component)
    expect(tree).toMatchSnapshot()
  })

  test('Should show level 1 title', () => {
    render(<Title level={1} />)
    expect(screen.getByRole('heading', { name: '', level: 1 })).toBeDefined()
  })

  test('Should show level 2 title', () => {
    render(<Title level={2} />)
    expect(screen.getByRole('heading', { name: '', level: 2 })).toBeDefined()
  })

  test('Should show level 3 title', () => {
    render(<Title level={3} />)
    expect(screen.getByRole('heading', { name: '', level: 3 })).toBeDefined()
  })

  test('Should show level 4 title', () => {
    render(<Title level={4} />)
    expect(screen.getByRole('heading', { name: '', level: 4 })).toBeDefined()
  })

  test('Should show default title', () => {
    const { container } = render(<Title />)
    expect(container.querySelector('div')).toBeDefined()
  })

  test('Should show children element', () => {
    const { container } = render(<Title>test</Title>)
    expect(container.querySelector('div')).toBeDefined()
    expect(screen.getByText('test')).toBeDefined()
  })
})
