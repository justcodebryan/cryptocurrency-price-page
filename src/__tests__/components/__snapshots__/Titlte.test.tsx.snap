// Vitest Snapshot v1

exports[`Title Test Suit > Title Snapshot 1`] = `
<div
  className="_title_35ba99"
/>
`;
